modules/afmongodb/afmongodb-legacy-uri.c

```diff
@@ -0,0 +1,301 @@
+/*
+ * Copyright (c) 2010-2016 Balabit
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "afmongodb-legacy-uri.h"
+#include "afmongodb-legacy-private.h"
+
+#define DEFAULTHOST "127.0.0.1"
+#define SOCKET_TIMEOUT_FOR_MONGO_CONNECTION_IN_MILLISECS 60000
+
+static gboolean
+_parse_addr(const char *str, char **host, gint *port)
+{
+  if (!host || !port)
+    {
+      msg_debug("Host or port reference should not be NULL");
+      return FALSE;
+    }
+  char *proto_str = g_strdup_printf("mongodb://%s", str);
+  mongoc_uri_t *uri = mongoc_uri_new(proto_str);
+  if (!uri)
+    {
+      msg_error("Cannot parse MongoDB URI", evt_tag_str("uri", proto_str));
+      g_free(proto_str);
+      return FALSE;
+    }
+
+  const mongoc_host_list_t *hosts = mongoc_uri_get_hosts(uri);
+  if (!hosts || hosts->next)
+    {
+      if (hosts)
+        msg_error("Multiple hosts found in MongoDB URI", evt_tag_str("uri", proto_str));
+      else
+        msg_error("No host found in MongoDB URI", evt_tag_str("uri", proto_str));
+      g_free(proto_str);
+      mongoc_uri_destroy(uri);
+      return FALSE;
+    }
+  *port = hosts->port;
+  *host = g_strdup(hosts->host);
+  mongoc_uri_destroy(uri);
+  if (!*host)
+    {
+      msg_error("NULL hostname", evt_tag_str("uri", proto_str));
+      g_free(proto_str);
+      return FALSE;
+    }
+  g_free(proto_str);
+  return TRUE;
+}
+
+static void
+_free_host_port(gpointer data)
+{
+  MongoDBHostPort *hp = (MongoDBHostPort *)data;
+  g_free(hp->host);
+  hp->host = NULL;
+  g_free(hp);
+}
+
+static gboolean
+_append_host(GList **list, const char *host, gint port)
+{
+  if (!list)
+    return FALSE;
+  MongoDBHostPort *hp = g_new0(MongoDBHostPort, 1);
+  hp->host = g_strdup(host);
```

Comment:
If we have a dedicated struct for host and port pairs, we might want to extract them to a separate constructor function. I am not sure we need mongodb legacy private.h, I would just merge it in if it didn't deserve a constructor either.